myapp/views.py

```diff
@@ -1,51 +1,83 @@
-	data = Stock.objects.filter(top_rank__isnull=False).order_by('top_rank')
-	return render(request, 'index.html', {'page_title': 'Main', 'data': data })
-
-
-# View for the single stock page
-# symbol is the requested stock's symbol ('AAPL' for Apple)
-def single_stock(request, symbol):
-	data = stock_api.get_stock_info(symbol)
-	return render(request, 'single_stock.html', {'page_title': 'Stock Page - %s' % symbol, 'data': data})
-
-
-def register(request):
-	# If post -> register the user and redirect to main page
-	if request.method == 'POST':
-		firstname = request.POST.get('firstname')
-		lastname = request.POST.get('lastname')
-		email = request.POST.get('email')
-		password = request.POST.get('password')
-
-		newuser = User.objects.create_user(username=email, email=email, password=password)
-		newuser.first_name = firstname
-		newuser.last_name = lastname
-		newuser.save()
-		return redirect('index')
-	else:
-		# If not post (regular request) -> render register page
-		return render(request, 'register.html', {'page_title': 'Register'})
-
-
-def logout_view(request):
-	logout(request)
-	return redirect('index')
-
-
-# API for a stock's price over time
-# symbol is the requested stock's symbol ('AAPL' for Apple)
-# The response is JSON data of an array composed of "snapshot" objects (date + stock info + ...), usually one per day
-def single_stock_historic(request, symbol):
-	data = stock_api.get_stock_historic_prices(symbol, time_range='1m')
-	return JsonResponse({'data': data})
+from django.shortcuts import render, redirect
+from myapp import stock_api
+from myapp.models import Stock
+from django.http import JsonResponse
+from django.contrib.auth.models import User
+from django.contrib.auth import logout
+from .exceptions.stock_service import StockServerUnReachable, StockSymbolNotFound
+
+
+# View for the home page - a list of 20 of the most active stocks
+def index(request):
+    # Query the stock table, filter for top ranked stocks and order by their rank.
+    data = Stock.objects.filter(top_rank__isnull=False).order_by('top_rank')
+    return render(request, 'index.html', {'page_title': 'Main', 'data': data})
+
+
+# View for the single stock page
+# symbol is the requested stock's symbol ('AAPL' for Apple)
+def single_stock(request, symbol):
+    try:
+        data = stock_api.get_stock_info(symbol)
+    except StockSymbolNotFound as e:
+        response = render(request, 'exception.html', {'404_error_message': e.message})
+        response.status_code = 404
+        return response
+    except StockServerUnReachable as e:
+        response = render(request, 'exception.html', {'503_error_message': e.message})
+        response.status_code = 503  # Service Unavailable code
+        return response
+    except Exception as e:
+        response = render(request, 'exception.html',
+                          {'404_error_message': "Unknown Error occurred: {}".format(", ".join(e.args))})
```

Comment:
In your context dictionary, change the error message key string to something other than `'404_error_message'`

Reply:
yes i already changed it in the branch `non-existend-symbol`, this part of the code doesn't belong to here. mistakenly merged changes from there to this branch, I will remove the merges and make it separate

Reply:
mmm, looks like there is some changes in `non-existent-symbol` needed here, it's really a problem to have branches counting on other branches didn't merge, causing little confusion. 
whatever, I merged the updated changes from there to this branch, now should be fine